File: languages/go/multi.go
package main

import (
	"encoding/binary"
	"fmt"
	"log"
	"math"
	"os"
	"sort"
	"strconv"
)

func lcgNext(s *uint32) uint32 {
	*s = 1664525*(*s) + 1013904223
	return *s
}

func sievePrimes(N int) {
	if N < 2 {
		fmt.Println(0)
		return
	}
	comp := make([]bool, N+1)
	comp[0], comp[1] = true, true
	lim := int(math.Sqrt(float64(N)))
	for p := 2; p <= lim; p++ {
		if !comp[p] {
			for m := p * p; m <= N; m += p {
				comp[m] = true
			}
		}
	}
	cnt := 0
	for i := 2; i <= N; i++ {
		if !comp[i] {
			cnt++
		}
	}
	fmt.Println(cnt)
}

func sortInts(N int) {
	arr := make([]uint32, N)
	s := uint32(123456789)
	for i := 0; i < N; i++ {
		arr[i] = lcgNext(&s)
	}
	sort.Slice(arr, func(i, j int) bool { return arr[i] < arr[j] })
	xorv := uint64(arr[0]) ^ uint64(arr[N/2]) ^ uint64(arr[N-1])
	var sum uint64 = 0
	for _, v := range arr {
		sum += uint64(v)
	}
	fmt.Printf("%d %d\n", xorv, sum)
}

func matmul(n int) {
	N := n
	A := make([]float64, N*N)
	B := make([]float64, N*N)
	C := make([]float64, N*N)
	s := uint32(123456789)
	inv := 1.0 / 4294967296.0
	for i := 0; i < N*N; i++ {
		A[i] = float64(lcgNext(&s)) * inv
	}
	for i := 0; i < N*N; i++ {
		B[i] = float64(lcgNext(&s)) * inv
	}
	for i := 0; i < N; i++ {
		row := i * N
		for k := 0; k < N; k++ {
			aik := A[row+k]
			col := k * N
			for j := 0; j < N; j++ {
				C[row+j] += aik * B[col+j]
			}
		}
	}
	sum := 0.0
	for i := 0; i < N*N; i++ {
		sum += C[i]
	}
	// print hex bit pattern
	u := math.Float64bits(sum)
	buf := make([]byte, 8)
	binary.BigEndian.PutUint64(buf, u)
	fmt.Printf("%016x\n", u)
}

func kmp(n, m int) {
	T := make([]byte, n)
	P := make([]byte, m)
	s := uint32(123456789)
	for i := 0; i < n; i++ {
		T[i] = byte('a' + (lcgNext(&s) % 26))
	}
	for i := 0; i < m; i++ {
		P[i] = byte('a' + (lcgNext(&s) % 26))
	}
	// build lps
	lps := make([]int, m)
	for i, l := 1, 0; i < m; {
		if P[i] == P[l] {
			l++
			lps[i] = l
			i++
		} else if l != 0 {
			l = lps[l-1]
		} else {
			lps[i] = 0
			i++
		}
	}
	// search
	count := 0
	for i, j := 0, 0; i < n; {
		if T[i] == P[j] {
			i++
			j++
			if j == m {
				count++
				j = lps[j-1]
			}
		} else if j != 0 {
			j = lps[j-1]
		} else {
			i++
		}
	}
	fmt.Println(count)
}

func main() {
	if len(os.Args) < 3 {
		log.Fatalf("usage: %s <task> <args...>", os.Args[0])
	}
	task := os.Args[1]
	switch task {
	case "sieve_primes":
		N, _ := strconv.Atoi(os.Args[2])
		sievePrimes(N)
	case "sort_ints":
		N, _ := strconv.Atoi(os.Args[2])
		sortInts(N)
	case "matmul_f64":
		n, _ := strconv.Atoi(os.Args[2])
		matmul(n)
	case "string_kmp":
		if len(os.Args) < 4 {
			log.Fatalf("need N M")
		}
		N, _ := strconv.Atoi(os.Args[2])
		M, _ := strconv.Atoi(os.Args[3])
		kmp(N, M)
	default:
		log.Fatalf("unknown task: %s", task)
	}
}

File: languages/lua/multi.lua
#!/usr/bin/env lua

local function lcg_next(s)
    return (s * 1664525 + 1013904223) & 0xFFFFFFFF
end

local function sieve_primes(N)
    if N < 2 then
        print(0)
        return
    end
    
    local comp = {}
    for i = 0, N do comp[i] = false end
    comp[0] = true
    comp[1] = true
    
    local lim = math.floor(math.sqrt(N))
    for p = 2, lim do
        if not comp[p] then
            local m = p * p
            while m <= N do
                comp[m] = true
                m = m + p
            end
        end
    end
    
    local cnt = 0
    for i = 2, N do
        if not comp[i] then
            cnt = cnt + 1
        end
    end
    print(cnt)
end

local function sort_ints(N)
    local s = 123456789
    local a = {}
    
    for i = 1, N do
        s = lcg_next(s)
        a[i] = s
    end
    
    table.sort(a)
    
    local xorv = (a[1] ~ a[math.floor(N/2) + 1] ~ a[N]) & 0xFFFFFFFF
    local total = 0
    for i = 1, N do
        total = (total + (a[i] & 0xFFFFFFFF)) & 0xFFFFFFFFFFFFFFFF
    end
    print(string.format("%u %u", xorv, total))
end

local function matmul(n)
    local N = n
    local s = 123456789
    local inv = 1.0 / 4294967296.0
    
    local A = {}
    local B = {}
    local C = {}
    
    for i = 1, N * N do
        s = lcg_next(s)
        A[i] = s * inv
    end
    for i = 1, N * N do
        s = lcg_next(s)
        B[i] = s * inv
    end
    for i = 1, N * N do
        C[i] = 0.0
    end
    
    for i = 0, N - 1 do
        local row = i * N
        for k = 0, N - 1 do
            local aik = A[row + k + 1]
            local col = k * N
            for j = 0, N - 1 do
                C[row + j + 1] = C[row + j + 1] + aik * B[col + j + 1]
            end
        end
    end
    
    local sm = 0.0
    for i = 1, N * N do
        sm = sm + C[i]
    end
    
    -- Convert float64 to hex representation
    local function double_to_hex(d)
        local s = string.pack(">d", d)
        return string.format("%016x", string.unpack(">I8", s))
    end
    
    print(double_to_hex(sm))
end

local function kmp(N, M)
    local s = 123456789
    local T = {}
    local P = {}
    
    for i = 1, N do
        s = lcg_next(s)
        T[i] = string.char(97 + (s % 26))
    end
    for i = 1, M do
        s = lcg_next(s)
        P[i] = string.char(97 + (s % 26))
    end
    
    local lps = {}
    for i = 1, M do lps[i] = 0 end
    
    local length = 0
    local i = 2
    while i <= M do
        if P[i] == P[length + 1] then
            length = length + 1
            lps[i] = length
            i = i + 1
        elseif length ~= 0 then
            length = lps[length]
        else
            lps[i] = 0
            i = i + 1
        end
    end
    
    local cnt = 0
    i = 1
    local j = 1
    while i <= N do
        if T[i] == P[j] then
            i = i + 1
            j = j + 1
            if j > M then
                cnt = cnt + 1
                j = lps[j - 1] + 1
            end
        elseif j > 1 then
            j = lps[j - 1] + 1
        else
            i = i + 1
        end
    end
    print(cnt)
end

local function main()
    if #arg < 2 then
        io.stderr:write("usage: multi.lua <task> <args...>\n")
        os.exit(1)
    end
    
    local task = arg[1]
    
    if task == "sieve_primes" then
        local N = tonumber(arg[2])
        sieve_primes(N)
    elseif task == "sort_ints" then
        local N = tonumber(arg[2])
        sort_ints(N)
    elseif task == "matmul_f64" then
        local n = tonumber(arg[2])
        matmul(n)
    elseif task == "string_kmp" then
        if #arg < 3 then
            io.stderr:write("need N M\n")
            os.exit(1)
        end
        local N = tonumber(arg[2])
        local M = tonumber(arg[3])
        kmp(N, M)
    else
        io.stderr:write("unknown task: " .. task .. "\n")
        os.exit(1)
    end
end

main()